active/0000-keyword-arguments.md

```diff
@@ -0,0 +1,139 @@
+```
+
+First of all, the call tells nothing the user about what those values mean.
+Secondly, their order is arbitrary and must be memorized. What I propose is that this call would rather look like:
+```rust
+window.addNewControl(
+    title => "Title", 
+    xPosition => 20, 
+    yPosition => 50, 
+    width => 100, 
+    height => 50, 
+    drawingNow => true);
+```
+
+While you might argue that this is more verbose, this is already the standard in the JavaScript ecosystem.
+A lot of libraries in JavaScript actually have a convention with calling with associative arrays like:
+```JavaScript
+window.addNewControl({ xPosition: 20, yPosition: 50, width: 100, height: 5,
+                 drawingNow: true });
+```
+If this wasn't a convenient pattern, nobody would bother to do it.
+
+An additional benefit is that this leads to an easy implementation of optional parameters and optional parameters.
+In languages like PHP, default parameters must come at the end of the function and must have a value.
+
+You can have optional parameters in the beginning, middle and end of functions without any default value.
+This means that this design allows for truly optional parameters without any run-time cost of using an `Option`.
+
+# Detailed design
+Currently Rust has
+
+```rust
+fn slice(&self, begin: usize, end: usize) -> &'a str
+fn slice_from(&self, begin: usize) -> &'a str
+fn slice_to(&self, end: usize) -> &'a str
+```
+
+This can be changed to
+
+```rust
+fn slice(&self, from => begin: usize, to => end: usize) -> &'a str
+fn slice(&self, from => begin: usize) -> &'a str
+fn slice(&self, to => end: usize) -> &'a str
+```
+
+Note that these are three different functions that have three different signatures.
+The keywords a function accepts is part of its signature. You can call these functions like this:
+
+```rust
+foo.slice(from => 5); //equivalent to current foo.slice_from(5)
+foo.slice(to => 9);   //equivalent to current foo.slice_to(9)
+foo.slice(from => 5, to => 9);       //equivalent to current foo.slice(5, 9)
+foo.slice(from => 9, to => 5);       //equivalent to current foo.slice(5, 9)
+```
+
+The trait looks like
+
+```rust
+Fn(&self, from => begin: usize, to => end: usize) -> &'a str
+//which desugars to
+Fn<({from: usize, to: usize}, &Self), &'a str>
+```
+
+So this is equivalent for using a `struct` to pass around your keyword parameters.
+But now you are able to use the same function name several times as long as the keywords are different.
+
+# Drawbacks
+
+This is a more complicated design than just having default arguments and overloading.
+Now there are both positional and keyword arguments and they interact with lifetimes, traits, and closures.
+
+# Alternatives
+
+A better design to the above function might be designing it like so:
+
+```rust
+let title = "Title";
+let position = Position(20, 50);
+let dimensions = Dimension(100, 50);
+window.addNewControlDrawingNow(title, position, dimensions);
+```
```

Comment:
This isn't *really* an alternative; both because it's something you can do right now *and* because you have to remember the order.  I think this falls under the traditional "alternative" of "don't change anything".